pkg/management/authorizationdiff.go

```diff
@@ -0,0 +1,136 @@
+}
+
+// An authorization set is a multiple-level map. It is created using a set of keys, each key is matching a level of the map.
+// While there are remaining keys, you add a level of maps: in this case, any is another map[string]any.
+// When all keys are processed, 'any' matches an instance of a struct, in our case, an instance of configuration.Authorization
+type authorizationSet struct {
+	authz map[string]any
+}
+
+func toNonNil(value *string) string {
+	if value == nil {
+		return nilString
+	}
+	return *value
+}
+
+// insertIntoMap will fill a multi-level map using given keys. Each key will be used at a different level.
+// When all keys have been used, the given authorization is stored.
+// Let's imagine we are not storing authorization but vehicles and let's consider the following calls:
+//
+//	insertIntoMap(myVehicles, Vehicle{Type: Car, Brand: Porsche, Color: Blue}, Car, Porsche, Blue)
+//	insertIntoMap(myVehicles, Vehicle{Type: Motorbike, Brand: Suzuki, Color: Red}, Motorbike, Suzuki, Red)
+//	insertIntoMap(myVehicles, Vehicle{Type: Car, Brand: Ferrari, Color: Red}, Car, Ferrari, Red)
+//
+// The map result is the following:
+// myVehicles
+//
+//	|
+//	+ Car
+//	|   |
+//	|   + Ferrari
+//	|   |   |
+//	|   |   + Red
+//	|   |       |
+//	|   |       + Vehicle{Type: Car, Brand: Ferrari, Color: Red}
+//	|   + Porsche
+//	|   |   |
+//	|   |   + Blue
+//	|   |       |
+//	|   |       + Vehicle{Type: Car, Brand: Porsche, Color: Blue}
+//	+ Motorbike
+//	|   |
+//	|   + Suzuki
+//	|       |
+//	|       + Red
+//	|           |
+//	|           + Vehicle{Type: Motorbike, Brand: Suzuki, Color: Red}
+func insertIntoMap(authz map[string]any, auth configuration.Authorization, keys ...string) {
+	var subMap any
+	if v, ok := authz[keys[0]]; ok {
+		// Key already exists... Get the matching sub map
+		subMap = v
+	} else {
+		// Key does not exist yet... Create a sub map
+		subMap = map[string]any{}
+	}
+	if len(keys) > 1 {
+		// There are remaining keys: insert auth in subMap using remaining keys (here, any is a map[string]any)
+		insertIntoMap(subMap.(map[string]any), auth, keys[1:]...)
+		authz[keys[0]] = subMap
+	} else {
+		// It was the last key: set the auth value (here, any is a configuration.Authorization)
+		authz[keys[0]] = auth
+	}
+}
+
+func toMap(input []configuration.Authorization) map[string]any {
+	var authz = make(map[string]any)
+	for _, auth := range input {
+		// Insert auth in the authz multi-level map using following different level of keys:
+		// - realm, group, action, targetRealm, targetGroup
+		insertIntoMap(authz, auth, *auth.RealmID, *auth.GroupName, *auth.Action, toNonNil(auth.TargetRealmID), toNonNil(auth.TargetGroupName))
+	}
+	return authz
+}
+
+func newAuthorizationSet(input []configuration.Authorization) AuthorizationSet {
+	return &authorizationSet{
+		authz: toMap(input),
+	}
+}
+
+func (d *authorizationSet) Apply(otherIntf AuthorizationSet, addFunction AlterFunction, removeFunction AlterFunction) {
```

Comment:
why not call it Diff instead of vague apply ?

Reply:
and instead of other, could we call other new ? because it is the new set of authorization. I think it could make easier to read

Reply:
Fixed